### ISSUE TITLE
helm-find-files unusable with Windows UNC path

### ISSUE THREAD

**Issue by neocryses** [emacs-helm/helm#2004]
## Expected behavior
With helm-find-files:
When typing file name such as foo.txt with UNC path (as shown below), texts typed in should appear as is.
**example UNC path:** //server/dir/foo.txt
## Actual behavior (from `emacs-helm.sh` if possible, see note at the bottom)
It is possible to type up to this part (cursor position is represented with a pipe)
```
//server/dir/foo|
```
but then if I type `.` right after it, it then becomes
```
//server/dir/foo./.|
```
pressing backspace once results in
```
//server/dir/foo./|
```
then pressing backspace again will revert back to the previous state (it's an infinite loop)
```
//server/dir/foo./.|
```
also, pressing C-l to Go up one level causes it to become (`/` at the beginning is gone)
```
/server/dir/|
```
## Steps to reproduce (recipe)
On Windows 10, using freshly downloaded Emacs (emacs-25.3_1-x86_64)

1. Fresh install helm (By adding melpa to package-archives, refreshing package-list, then installing via package.el)
1. Require it
1. `M-x helm-find-files`
1. Try to open a file in a UNC path by typing in the path. e.g. `//server/dir/foo.txt`

Same behaviour can be observed by open a file in UNC path then trying to delete the file extension.
To observe C-l's behaviour, open a file in UNC path then press C-l.
## Describe versions of Helm, Emacs, operating system, etc.
helm-20180326.109
GNU Emacs 25.3.1 (x86_64-mingw32) of 2017-09-27
Windows 10, Windows 7 and macOS High Sierra
## Are you using `emacs-helm.sh` to reproduce this bug (yes/no):
no for Windows 10 & 7 (not possible natively right)
yes for macOS

I assume this is caused by automatic expansion of `//` to `/` or `C:` defined in helm-file.el around [here](https://github.com/emacs-helm/helm/blob/36e620c4d9cbb29d7cba97ccce736792192f981e/helm-files.el#L1944-L1951)
Couldn't figure out the cause. I'm new to emacs-lisp.

This doesn't seem to happen with the stock `M-x find-file` command, or ivy's `M-x counsel-find-file`. Ivy also expands `//` to `/`, but it seems to handle UNC path without any problem.

I know I can get around this by mounting the samba shares, but since I'm forced to work using Windows and everyone uses Windows there, I often get sent UNC path to open files from and mounting each one isn't really a solution.

I tried Ivy but it lacks some of the functionality from helm I cannot let go of. I'd be very thankful if you can fix this issue.

**Comment by thierryvolpiatto** [comment]
Neocryses <notifications@github.com> writes:

> Expected behavior
>
> With helm-find-files:
> When typing file name such as foo.txt with UNC path (as shown below), texts typed in should appear as is.
> example UNC path: //server/dir/foo.txt

I have no Windows installation under hand so I am afraid I will have to
fix this blindly with your help if possible.
First try, is evaluating code below fix your problem?

    (defun helm-ff--expand-file-name-no-dot (name &optional directory)
      (expand-file-name name directory))

-- 
Thierry

**Comment by neocryses** [comment]
Thanks for taking your time on this. I will do my best to assist you.

Evaluating that code does indeed fix the problem of typing `foo.` becoming `foo./.`.
Trying to delete the dot from `foo.` is also not a problem after evaluating the code.

However, the behaviour of C-l is still there. Pressing C-l after typing `//server/dir/foo.txt` causes it to become `/server/dir/`.

Regarding the behavior of C-l, I have googled about this before and did find a workaround in a Japanese NTEmacs wiki [here](https://www49.atwiki.jp/ntemacs/pages/32.html).
Below is the snippet from the wiki:
```
(advice-add 'helm-reduce-file-name
            :override (lambda (&rest args)
                        (let ((fname (nth 0 args))
                              (level (nth 1 args)))
                          (while (> level 0)
                            (setq fname (expand-file-name (concat fname "/../")))
                            (setq level (1- level)))
                          fname)))
```

Evaluating above code together with the code you provided seems to fix the issue.

**Comment by thierryvolpiatto** [comment]
Neocryses <notifications@github.com> writes:

> Thanks for taking your time on this. I will do my best to assist you.

Thanks.

> Evaluating that code does indeed fix the problem of typing
> foo. becoming foo./..

Ok so I will make a first commit fixing this issue.

> However, the behaviour of C-l is still there. Pressing C-l after
> typing //server/dir/foo.txt causes it to become /server/dir/.
>
> Regarding the behavior of C-l, I have googled about this before and
> did find a workaround in a Japanese NTEmacs wiki here.  Below is the
> snippet from the wiki:
>
> (advice-add 'helm-reduce-file-name
>             :override (lambda (&rest args)
>                         (let ((fname (nth 0 args))
>                               (level (nth 1 args)))
>                           (while (> level 0)
>                             (setq fname (expand-file-name (concat fname "/../")))
>                             (setq level (1- level)))
>                           fname)))
>
> Evaluating above code together with the code you provided seems to fix the issue.

Ok, so can you give me the output of 

`(helm-reduce-file-name "//server/dir/foo.txt" 1)`

before applying the advice and after.

Thanks.

--
Thierry

**Comment by thierryvolpiatto** [comment]
Neocryses <notifications@github.com> writes:

> Here are the results:
> Evaluating helm-reduce-file-name "//server/dir/foo.txt" 1) before applying the advice results in
>
> "/server/dir/"
>
> and after applying the advice
>
> "//server/dir/"

Ok, so let's use the version you provided, it seems to work fine on
Linux as well.

> Thanks for your swift action on the first issue.
> I tested 7214729 and it worked flawlessly.

Great.

-- 
Thierry

**Comment by neocryses** [comment]
Just tested both 72147290fbded28f903a1fa52f9184aaded3df89 and 5b95ea908a85ce92f87104d46567d4c7a80d3138 on Windows 10, and both of the issue seems to have been resolved. Thank you very much for the work!

Though I have some things I'd like to clarify, just for the sake of completeness.

### First question
Since `helm-ff--expand-substitued-pattern` does a check to see if it's run on Windows, trying to type in "//server/dir/foo.txt" in non-Windows OS still has the issue of `.` becoming `./.`.
Is this a side-effect of a feature specific to non-Windows OS?

Even though the UNC path convention of double-slash at the beginning of the line isn't supported for non-Windows OS, I see no reason to leave this behavior unless the cause of this issue is a side-effect of another feature.

### Second question
Pressing C-l when the minibuffer content is `//server/` does nothing.
I guess it makes sense because `//server/` is the root of the remote server, just like how `C:` is the root equivalent for Windows systems.
I'm just wondering if this behavior is intended or not. If yes, that's totally fine. If no, that's totally fine too!

**Comment by thierryvolpiatto** [comment]
Neocryses <notifications@github.com> writes:

> Just tested both 7214729 and 5b95ea9 on Windows 10, and both of the issue seems to have been resolved. Thank you very much for the work!
>
> Though I have some things I'd like to clarify, just for the sake of completeness.
>
> First question
>
> Since helm-ff--expand-substitued-pattern does a check to see if it's run on Windows, trying to type in "//server/dir/foo.txt" in non-Windows OS still has the issue of
> . becoming ./..
> Is this a side-effect of a feature specific to non-Windows OS?
>
> Even though the UNC path convention of double-slash at the beginning of the line isn't supported for non-Windows OS, I see no reason to leave this behavior unless the
> cause of this issue is a side-effect of another feature.
>
> Second question
>
> Pressing C-l when the minibuffer content is //server/ does nothing.
> I guess it makes sense because //server/ is the root of the remote server, just like how C: is the root equivalent for Windows systems.
> I'm just wondering if this behavior is intended or not. If yes, that's totally fine. If no, that's totally fine too!

No it is not intended, i didn't think at this thanks to catch this, I
will try to fix it ASAP.

-- 
Thierry

**Comment by thierryvolpiatto** [comment]
Neocryses <notifications@github.com> writes:

> Just tested both 7214729 and 5b95ea9 on Windows 10, and both of the issue seems to have been resolved. Thank you very much for the work!
>
> Though I have some things I'd like to clarify, just for the sake of completeness.
>
> First question
>
> Since helm-ff--expand-substitued-pattern does a check to see if it's
> run on Windows, trying to type in "//server/dir/foo.txt" in
> non-Windows OS still has the issue of . becoming ./..  Is this a
> side-effect of a feature specific to non-Windows OS?
>
> Even though the UNC path convention of double-slash at the beginning
> of the line isn't supported for non-Windows OS, I see no reason to
> leave this behavior unless the cause of this issue is a side-effect of
> another feature.

I have commited another fix that prevent adding the second / on Linux,
so this issue should be fixed, thanks to catch this, it is not new or
relative to recent changes, looks like a very old bug nobody noticed
until now!!!

> Second question
>
> Pressing C-l when the minibuffer content is //server/ does nothing.  I
> guess it makes sense because //server/ is the root of the remote
> server, just like how C: is the root equivalent for Windows systems.
> I'm just wondering if this behavior is intended or not. If yes, that's
> totally fine. If no, that's totally fine too!

Well yes really unintended as I know nothing of UNC paths and like but
if hopefully it works...that's fine yes!

Let me know if you find other issues and thanks for your help on this.

--
Thierry

**Comment by neocryses** [comment]
I have applied the fix 72147290fbded28f903a1fa52f9184aaded3df89, 5b95ea908a85ce92f87104d46567d4c7a80d3138, 6a34d57f416e4194e6d3207b558d3a00cdf2b955 on Windows 10 and macOS High Sierra, and both seems to work perfectly as intended, including the prevention of double-slash in non-Windows OS!

> Well yes really unintended as I know nothing of UNC paths and like but
> if hopefully it works...that's fine yes!

Some may have different opinion on this behaviour, but from what I know of UNC path, I reckon this is a reasonable behaviour. I see no other side-effects with this either, so it should be fine.

I will close the issue because all of my problems are solved! I will let you know if I find any issue in the future. Hopefully by then I'll have some emacs-lisp knowledge to be of more help to you.

Thanks so much for your help on this! Much appreciated.
